[MIG] purchase_analytic: Migration to 17.0

CLaurelB opened the pull request:
Supersed https://github.com/OCA/account-analytic/pull/707

Commit:
module to manage analytic on purchase order

Commit:
[10.0][MIG] port purchase_analytic to V10.0

oca-travis committed:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[ADD] icon.png

MiquelRForgeFlow committed:
[MIG] purchase_analytic: Migration to 12.0

oca-travis committed:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-12.0/account-analytic-12.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-12-0/account-analytic-12-0-purchase_analytic/

ps-tubtim committed:
[IMP] purchase_analytic: black, isort

ps-tubtim committed:
[MIG] purchase_analytic: Migration to 13.0

oca-travis committed:
[UPD] Update purchase_analytic.pot

eduardgm committed:
Added translation using Weblate (Spanish)

alan196 committed:
[IMP] purchase_analytic: black, isort, prettier

alan196 committed:
[MIG] purchase_analytic: Migration to 15.0

Commit:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[UPD] README.rst

LoisRForgeFlow committed:
[IMP] purchase_analytic: adapt to newest framework.

Some of the "tricks" done in this module are no longer needed
and can be easily implemented with newest framework features:

* No need for an auxiliar `project_id2` field. User can set
  an analytic account with no lines and it is respected.
* Simplify onchange. Now update analytic line on the go (no
  need to save) which is a better UX because avoid unexpected
  changes on save.

Also re-label the field `project_id` to "Analytic Account" to
align with the typical label in newer versions of Odoo.
Tha label "Contract / Analytic" was last used in v8 (
https://github.com/odoo/odoo/blob/8.0/addons/sale/sale.py#L217).

Commit:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
purchase_analytic 15.0.2.0.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-15.0/account-analytic-15.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-15-0/account-analytic-15-0-purchase_analytic/

AungKoKoLin1997 committed:
[MIG] purchase_analytic: Migration to 16.0

AungKoKoLin1997 committed:
analytic_distribution readonly states

Commit:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[UPD] README.rst

badbole committed:
Added translation using Weblate (Croatian)

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (3 of 3 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/hr/

Gelo-fl committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (3 of 3 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/es/

sanrav committed:
Translated using Weblate (French)

Currently translated at 100.0% (3 of 3 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/fr/

Commit:
[UPD] Update purchase_analytic.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/

sanrav committed:
Translated using Weblate (French)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/fr/

sanrav committed:
Added translation using Weblate (French (France) (fr_FR))

sanrav committed:
Translated using Weblate (French (France) (fr_FR))

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/fr_FR/

AungKoKoLin1997 committed:
[IMP] purchase_analytic: add default value in order line

OCA-git-bot committed:
purchase_analytic 16.0.2.0.0

adrianojprado committed:
Added translation using Weblate (Portuguese (Brazil))

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/pt_BR/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/es/

OCA-git-bot committed:
[UPD] README.rst

AungKoKoLin1997 committed:
[FIX] purchase_analytic: Access Denied by ACLs

OCA-git-bot committed:
[BOT] post-merge updates

sbidoul committed:
[TRN] remove fr_FR.po

Co-authored-by: Alexis de Lattre <alexis@via.ecp.fr>

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/it/

Commit:
[UPD] Update purchase_analytic.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (5 of 5 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/it/

Commit:
[UPD] Update purchase_analytic.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (5 of 5 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/it/

SottomaiorMacedoTec committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (5 of 5 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/pt_BR/

Commit:
[UPD] Update purchase_analytic.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (7 of 7 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/it/

SottomaiorMacedoTec committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (7 of 7 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/pt_BR/

CLaurelB committed:
[IMP] purchase_analytic: pre-commit auto fixes

CLaurelB commented:
Hello @luisg123v @StefanRijnhart 

Could you please review this PR?

It already includes the additional tests to increase the coverage.

Best regards.

luisg123v commented on purchase_analytic/models/purchase.py:
```diff
@@ -30,14 +26,17 @@ def _compute_analytic_distribution(self):
     def _inverse_analytic_distribution(self):
-        """When set analytic_distribution set analytic distribution on all order lines"""
+        """When setting analytic_distribution, apply the analytic distribution to all
+        order lines."""
```
Nitpick:

Move closing quotes to the next lines, to follow PEP8 recommendations regarding multi-line docstrings

CLaurelB replied:
Updated.

luisg123v commented on purchase_analytic/views/purchase_views.xml:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8" ?>
+<?xml version="1.0" ?>
```
Why? I'd avoid this, AFAIK it's preferred to provide explicitly the encoding.

CLaurelB replied:
Change reverted.

luisg123v commented on purchase_analytic/tests/test_purchase_analytic.py:
```diff
@@ -17,65 +18,74 @@ def setUp(self):
+        # Test setting analytic distribution with an order line
+        po.order_line = self.add_new_order_line()
+        po.analytic_distribution = self.analytic_distribution_manual2
+        po._onchange_analytic_distribution()
```
If you need to cover an onchange, it's preferred to use the `Form` class

CLaurelB replied:
Refactored to use the `Form` class.

luisg123v commented on purchase_analytic/tests/test_purchase_analytic.py:
```diff
@@ -17,65 +18,74 @@ def setUp(self):
+                    "product_id": self.product_id.id,
+                    "product_qty": 1.0,
+                    "product_uom": self.uom_id.id,
+                    "price_unit": 121.0,
```
I'd suggest making these params configurable and provide default values, so it may be called without params but also being able to provide some of them.

CLaurelB replied:
As part of the refactoring, the non-required fields were removed, and the product was set as a parameter.

luisg123v commented on purchase_analytic/tests/test_purchase_analytic.py:
```diff
@@ -17,65 +18,74 @@ def setUp(self):
-    def test_analytic_distribution(self):
+    def add_new_order_line(self):
+        return [
+            Command.create(
```
Why don't you pass the order as param?

CLaurelB replied:
Done.

luisg123v commented on purchase_analytic/tests/test_purchase_analytic.py:
```diff
@@ -17,65 +18,74 @@ def setUp(self):
-                ],
-            }
-        )
+        po = self.env["purchase.order"].new({"partner_id": self.partner_id.id})
```
Te `Form` class?

CLaurelB replied:
Done.

luisg123v commented on purchase_analytic/tests/test_purchase_analytic.py:
```diff
@@ -12,9 +12,7 @@ def setUp(self):
-        analytic_plan = self.env["account.analytic.plan"].create(
-            {"name": "Plan Test", "company_id": False}
-        )
+        analytic_plan = self.env["account.analytic.plan"].create({"name": "Plan Test"})
```
Why this? Document it in commit description.

CLaurelB replied:
Documented in the commit.

CLaurelB committed:
[MIG] purchase_analytic: Migration to 17.0

Changelog:
- Set the `readonly` attribute of the purchase analytic field on the
  views based on its state, instead of defining it in the model.
- Update method docstrings to be more descriptive.
- Refactor tests to use `SetUpClass` and remove the unnecessary `_id`
  when acquiring a record (not just the id).
- Remove the unrequired parameters when creating a line.
- Add test cases for setting the analytic distribution on a purchase
  order without lines and unsetting the analytic distribution on a
  purchase order with lines.
- Use `Form` to test the onchange method.
- Add the user to the "Analytic Accounting" group in order to visualize
  analytic distributions in the forms.
- Remove the `company_id` field from the creation of `analytic_plan` as
  the field was moved to the `account.analytic.applicability` model in
  [1].

[1]: https://github.com/odoo/odoo/commit/dc696c8e

Co-authored-by: Wodran Van de Sande <wodran.vandesande@dynapps.be>

CLaurelB commented:
Closing because the original PR is active again.

CLaurelB commented:
Reopening this PR because @aisopuro may not have time to address the issues raised during the review here: https://github.com/OCA/account-analytic/pull/707

luisg123v reviewed (approved):
LGTM :+1: 

@moylop260 could you review/merge, please?

Regards,

StefanRijnhart commented on purchase_analytic/models/purchase.py:
```diff
@@ -30,14 +26,18 @@ def _compute_analytic_distribution(self):
     def _inverse_analytic_distribution(self):
-        """When set analytic_distribution set analytic distribution on all order lines"""
+        """When setting analytic_distribution, apply the analytic distribution to all
+        order lines.
```
The summary of a docstring should fit on one line, see https://peps.python.org/pep-0257/#multi-line-docstrings. Would you mind trying to make it fit, or to set a summary and additional info after the empty line?

CLaurelB replied:
Updated in all cases.

StefanRijnhart commented on purchase_analytic/models/purchase.py:
```diff
@@ -30,14 +26,18 @@ def _compute_analytic_distribution(self):
     def _onchange_analytic_distribution(self):
-        """When change analytic_distribution set analytic distribution on all order lines"""
+        """When changing analytic_distribution, apply the analytic distribution to all
+        order lines.
```
Same

CLaurelB replied:
Updated in all cases.

StefanRijnhart reviewed (commented):
Nice work! I have small request though.

StefanRijnhart reviewed (approved):
Thanks!

StefanRijnhart commented:
/ocabot migration purchase_analytic

StefanRijnhart commented:
/ocabot merge nobump
